src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -1290,7 +1290,7 @@ private void ThrowOnInvalidHandle(SafeWinHttpHandle handle)
         
         private Task<bool> InternalSendRequestAsync(WinHttpRequestState state)
         {
-            state.TcsSendRequest = new TaskCompletionSource<bool>();
+            state.TcsSendRequest = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
```

Comment:
The use of blocking Stream APIs is something we can't control very well and those calls are in library code that the developer can't really control.  So, using the RunContinuationsAsynchronously for the tasks solves that problem.

In terms of the certificate callback, I still think it is reasonable to document to say that it shouldn't block.  It's a new API and we should not impair performance of WinHttpHandler by omitting the WINHTTP_OPTION_ASSURED_NON_BLOCKING_CALLBACKS option just because someone is using a certificate validation callback. It is does block, then we have the same problem which results in hangs. But we educate developers about that.

It's possible we might change our mind on this but I would wait for customer feedback on that.